soc/xtensa/intel_adsp/common/include/cavs_hda.h

```diff
@@ -0,0 +1,303 @@
+
+/* Gateway Buffer Read Pointer */
+#define DGBRP(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x10))
+
+/* Gateway Buffer Write Pointer */
+#define DGBWP(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x14))
+
+/* Gateway Buffer Segment Pointer */
+#define DGBSP(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x18))
+
+/* Gateway Minimum Buffer Size */
+#define DGMBS(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x1c))
+
+/* Gateway Linear Link Position Increment */
+#define DGLLPI(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x24))
+
+/* Gateway Linear Position In Buffer Increment */
+#define DGLPIBI(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x28))
+
+/**
+ * @brief Dump all the useful registers of an HDA stream to printk
+ *
+ * This can be invaluable when finding out why HDA isn't doing (or maybe is)
+ * doing what you want it to do. Macro so you get the file and line
+ * of the call site included.
+ *
+ * @param name String that contains a name of the hda stream (or anything really)
+ * @param base Base address of the IP register block
+ * @param sid Stream ID
+ */
+#define cavs_hda_dbg(name, base, sid)					\
+		printk("%s:%u %s(%u:0x%p), dgcs: 0x%x, dgbba 0x%x, "	\
+		       "dgbs %u, dgbrp %u, dgbwp %u, dgbsp %u, "	\
+		       "dgmbs %u, dgbllpi 0x%x, dglpibi 0x%x\n",	\
+		       __FILE__, __LINE__, name,			\
+		       sid, DGCS(base, sid),				\
+		       *DGCS(base, sid),				\
+		       *DGBBA(base, sid),				\
+		       *DGBS(base, sid),				\
+		       *DGBRP(base, sid),				\
+		       *DGBWP(base, sid),				\
+		       *DGBSP(base, sid),				\
+		       *DGMBS(base, sid),				\
+		       *DGLLPI(base, sid),				\
+		       *DGLPIBI(base, sid))
+
+
+
+/**
+ * @brief Initialize an HDA stream for use with the firmware
+ *
+ * @param hda Stream set to work with
+ * @param sid Stream ID
+ */
+static inline void cavs_hda_init(uint32_t base, uint32_t sid)
+{
+	*DGCS(base, sid) |= DGCS_FWCB;
+}
+
+/**
+ * @brief Set the buffer, size, and element size for an HDA stream
+ *
+ * Sanity checks that the buffer address and size are valid and that the
+ * stream isn't enabled or busy.
+ *
+ * Prior to enabling an HDA stream to/from the host this is the minimum configuration
+ * that is required. It must be set *after* the host has configured its own buffers.
+ *
+ *
+ * @param hda Stream set to work with
+ * @param sid Stream ID
+ * @param buf Buffer address to use for the shared FIFO. Must be in L2 and 128 byte aligned.
+ * @param buf_size Buffer size in bytes Must be 128 byte aligned
+ *
+ * @retval -EBUSY if the HDA stream is already enabled
+ * @retval -EINVAL if the buf is not in L2, buf isn't aligned on 128 byte boundaries
+ * @retval 0 on Success
+ */
+static inline int cavs_hda_set_buffer(uint32_t base, uint32_t sid,
+				     uint8_t *buf, uint32_t buf_size)
+{
+	void *cached_buf = arch_xtensa_cached_ptr(buf);
```

Comment:
I'm confused here: why does the address need to be in the cached mapping?  Does the hardware require this?  Per spec, the DMA layer should be upstream of the xtensa mapping and shouldn't care, no?   If it doesn't work to use the uncached address, we should have a comment here calling that out.

Reply:
Its done so I can check the address is in L2  space which the hardware asks for. My interpretation may be wrong of the hardware docs but it does say cached or local L2. Local L2, I believe, is a different thing. Would the address range specified by cavs-mem L2_SRAM_BASE and L2_SRAM_SIZE be valid still if I chose uncached? That's my intent here, to ensure the address is in the range needed by the hardware to work.

Reply:
It's a 24 bit field which is fine, the goal in using the cached address is to validate the address is within the specified range and aligned appropriately. Fixed